audio-worklet/src/noisegate-script-processor.js

```diff
@@ -0,0 +1,163 @@
+    if (options == null) options = {};
+    this.context_ = context;
+    let channels = options.channels || 1;
+    let bufferSize = options.bufferSize || 0;
+    this.threshold = options.threshold || 0;
+    this.attack = options.attack || 0;
+    this.release = options.release || 0;
+    // Alpha controls a tradeoff between the smoothness of the 
+    // envelope and its delay, with a higher value giving more smoothness at 
+    // the expense of delay and vice versa. It should be set experimentally to
+    // control this tradeoff. 0.90 has proven an effective default in offline
+    // observations.
+    this.alpha_ = options.alpha || 0.90;
+    if (channels > 2)
+      throw 'The maximum supported channels is two.';
+
+    this.node_ = this.context_.createScriptProcessor(
+        bufferSize, channels, channels);
+    this.node_.onaudioprocess = this.onaudioprocess_.bind(this);
+    // The noise gate is connected to and from by dummy input and output nodes.
+    this.input = new GainNode(this.context_);
+    this.output = new GainNode(this.context_);
+    this.input.connect(this.node_).connect(this.output);
+    // The last envelope level of a given buffer is initialized as 0 
+    // for between-block memory.
+    this.lastLevel = 0;
+    // The last weight (between 0 and 1) assigned, where 1 means the gate 
+    // is open and 0 means it is closed and the sample in the output buffer is
+    // muted.
+    this.lastWeight = 1;
+  }
+
+  /**
+   * Gradually mutes input which registers beneath specified threshold.
+   * @param  {AudioProcessingEvent} event input and output buffers
+   */
+  onaudioprocess_(event) {
+    let envelope = this.detectLevel(event.inputBuffer);
+
+    for (let i = 0; i < event.inputBuffer.numberOfChannels; i++) {
+      let input = event.inputBuffer.getChannelData(i);
+      let output = event.outputBuffer.getChannelData(i);
+      let weights = this.computeGain(envelope);
+
+      for (let j = 0; j < input.length; j++) {
+        output[j] = weights[j] * input[j];
+      }
+    }
+  }
+
+  /**
+   * Detect level using the difference equation for the Root Mean Squared 
+   * value of the input signal.
+   * @param  {AudioBuffer} inputBuffer input audio buffer
+   * @return {Float32Array} envelope the registered level of the input             
+   */
+  detectLevel(inputBuffer) {
+    let channel = new Float32Array(event.inputBuffer.getChannelData(0).length);
+    // Stereo input is downmixed to mono input via averaging.
+    if (channel.numberOfChannels == 2) {
+      for (let i = 0; i < inputBuffer.getChannelData(0).length; i++) {
+        channel[i] = (inputBuffer.getChannelData(0)[i] +
+                          inputBuffer.getChannelData(1)[i]) / 2;
+      }
+    } else {
+      channel = inputBuffer.getChannelData(0);
+    }
+    // Set the value at each index of the envelope to be a positive value
+    // that combines the current and previous value according to alpha.
```

Comment:
This isn't really helpful since it describes very loosely exactly what the code is doing.  This is an exponential smoothing filter. Or leaky integrator. Or something else.